docs/security.rst

```diff
@@ -0,0 +1,232 @@
+    decompressed = decompress(decrypted)
+
+    ASSERT( CONSTANT_TIME_COMPARISON( chunk-id, AUTHENTICATOR(id_key, decompressed) ) )
+
+Offline key security
+--------------------
+
+Borg cannot secure the key material while it is running, because the keys
+are needed in plain to decrypt/encrypt repository objects.
+
+For offline storage of the encryption keys they are encrypted with a
+user-chosen passphrase. A 256 bit key encryption key (KEK) is derived
+from the passphrase using PBKDF2-HMAC-SHA256 with a random 256 bit
+salt which is then used to Encrypt-then-MAC a packed
+representation of the keys with AES-256-CTR with a constant
+initialization vector of 0. The resulting MAC is stored alongside the
+ciphertext and converted to base64.
+
+This base64 blob (commonly referred to as *keyblob*)
+is then stored in the key file or in the repository (keyfile and
+repokey modes respectively).
+
+This scheme, and specifically the use of a constant IV with the CTR
+mode, is secure because an identical passphrase will result in a
+different derived KEK for every encryption due to the salt.
+
+.. _Horton principle: https://en.wikipedia.org/wiki/Horton_Principle
+.. _HKDF: https://tools.ietf.org/html/rfc5869
+.. _length extension: https://en.wikipedia.org/wiki/Length_extension_attack
+
+Remote RPC protocol security
+============================
+
+.. note:: This section could use expansion.
+
+The RPC protocol is fundamentally based on msgpack'd messages exchanged
+over an encrypted SSH channel (the system's SSH client is used for this
+by piping data from/to it).
+
+This means that the authorization and transport security properties
+are inherited from SSH and the configuration of the SSH client
+and the SSH server. Therefore the remainder of this section
+will focus on the security of the RPC protocol within Borg.
+
+The assumed worst-case a server can inflict to a client is a
+denial of repository service.
+
+The situation were a server can
+create a general DoS on the client should be avoided, but
+might be possible by eg. forcing the client to allocate large
+amounts of memory to decode large messages. However, since we
+are not aware of any `"zip bomb"`_ equivalent in msgpack this
+would mean that the server has to send a similar amount of data to
+the client, which is not always practical. This is tracked with :issue:`2139`.
+
+We believe that other kinds of attacks, especially critical vulnerabilities
+like remote code execution are inhibited by the design of the protocol:
+
+1. The server cannot send requests to the client on it's own accord,
+   it only can send responses. This avoids "unexpected inversion of control"
+   issues.
+2. msgpack serialization does not allow embedding or referencing code that
+   is automatically executed. Incoming messages are unpacked by the msgpack
+   unpacker into native Python data structures (like tuples and dictionaries),
+   which are then passed to the rest of the program.
```

Comment:
i think the msgpack stuff is good enough for us, if someone is interested in msgpack, he can specifically look at that project. like he can look at ssh for ssh security properties...